dashboard/test/models/concerns/user/level_progressable_test.rb

```diff
@@ -0,0 +1,671 @@
+require 'test_helper'
+
+class LevelProgressableTest < ActiveSupport::TestCase
+  include Minitest::RSpecMocks
+
+  def create_level_group(sub_level_name)
+    level_group_dsl = <<~DSL
+      name 'LevelGroupLevel1'
+
+      page
+      level '#{sub_level_name}'
+    DSL
+    LevelGroup.create_from_level_builder({}, {name: 'LevelGroupLevel1', dsl_text: level_group_dsl})
+  end
+
+  def put_participant_in_section(participant, instructor, script, unit_group = nil, participant_type = 'student')
+    section = create :section, user_id: instructor.id, script_id: script.try(:id), course_id: unit_group.try(:id), participant_type: participant_type, grades: participant_type == 'student' ? ['9'] : ['pl']
+    Follower.create!(section_id: section.id, student_user_id: participant.id, user: instructor)
+    section
+  end
+
+  describe '#next_unpassed_visible_progression_level' do
+    subject {user.next_unpassed_visible_progression_level(script)}
+
+    context 'when none of the lessons are hidden' do
+      let(:user) {create(:user)}
+      let(:script) {create(:script, :with_levels, levels_count: 5)}
+
+      context 'when the user has no progress' do
+        it 'returns the first visible level' do
+          _(subject.chapter).must_equal 1
+        end
+      end
+
+      context 'when the user has some progress' do
+        let(:second_script_level) {script.get_script_level_by_chapter(2)}
+
+        before do
+          UserLevel.create(
+            user: user,
+            level: second_script_level.level,
+            script: script,
+            attempts: 1,
+            best_result: Activity::MINIMUM_PASS_RESULT
+          )
+        end
+        it 'returns the next visible level' do
+          _(subject.chapter).must_equal 3
+        end
+      end
+
+      context 'when the user skips a level' do
+        before do
+          script.script_levels.take(3).each do |sl|
+            UserLevel.create(
+              user: user,
+              level: sl.level,
+              script: script,
+              attempts: 1,
+              best_result: Activity::MINIMUM_PASS_RESULT
+            )
+          end
+          UserLevel.create(
+            user: user,
+            level: script.script_levels.last.level,
+            script: script,
+            attempts: 1,
+            best_result: Activity::MINIMUM_PASS_RESULT
+          )
+        end
+
+        it 'returns the skipped level' do
+          _(subject.chapter).must_equal 4
+        end
+      end
```

Comment:
Looks like we combined the tests `can get next_unpassed_visible_progression_level, last level complete, but script not complete, none hidden` and `can get next_unpassed_visible_progression_level, user skips level, none hidden`; is that right? I think that makes sense to do, just wanted to confirm my understanding

Reply:
Yes, some of the scenarios were testing the same behaviour. So I consolidate them into one test case.